### PR TITLE
Keep output as scalar for scalar function if all inputs are scalar

### LINKED ISSUE #7959: Error evaluating manually constructed physical expression: `<column> = left(<literal>, 1)`
### Describe the bug

When passed to `create_physical_expr`, the following expression works as expected on DataFusion 32.0.0 but fails on the `master` branch: 

```rust
col("column").eq(left("value".lit(), 1i64.lit()))
```

The following error appears:
```rust
Error: ArrowError(InvalidArgumentError("Cannot compare arrays of different lengths, got 4 vs 1"))
```

### To Reproduce

Compile the following piece of Rust code against the current master branch of DataFusion:

```rust
use std::sync::Arc;

use arrow::{array::StringArray, record_batch::RecordBatch};
use arrow_schema::{DataType, Field, Schema};
use datafusion::physical_expr::{create_physical_expr, execution_props::ExecutionProps};
use datafusion_common::{DFSchema, Result};
use datafusion_expr::{col, left, Literal};

#[tokio::main]
async fn main() -> Result<()> {
    let expr = col("letter").eq(left("AAPL".lit(), 1i64.lit()));
    println!("{expr}");

    let schema = Schema::new(vec![Field::new("letter", DataType::Utf8, false)]);
    let df_schema = DFSchema::try_from_qualified_schema("data", &schema)?;
    let p = create_physical_expr(&expr, &df_schema, &schema, &ExecutionProps::new())?;
    println!("{p}");

    let batch = RecordBatch::try_new(
        Arc::new(schema),
        vec![Arc::new(StringArray::from_iter_values(vec![
            "A", "B", "C", "D",
        ]))],
    )?;
    let z = p.evaluate(&batch)?;
    println!("{z:?}");

    Ok(())
}

// letter = left(Utf8("AAPL"), Int64(1))
// letter@0 = left(AAPL, 1)
// Error: ArrowError(InvalidArgumentError("Cannot compare arrays of different lengths, got 4 vs 1"))
```

### Expected behavior

`create_physical_expr` should produce a valid physical expression from the provided logical expression

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by viirya**
## Which issue does this PR close?

Closes #7959.

### CODE REVIEW

File: datafusion/sqllogictest/test_files/scalar.slt [hunk 1]
```diff
@@ -1878,3 +1878,51 @@ query T
...
+  ('B', 'APACHE'),
+  ('C', 'APACHE'),
+  ('D', 'APACHE')
+;
+
+query TT
+EXPLAIN SELECT letter, letter = LEFT('APACHE', 1) FROM simple_string;
+----
+logical_plan
+Projection: simple_string.letter, simple_string.letter = Utf8("A") AS simple_string.letter = left(Utf8("APACHE"),Int64(1))
```
Comment (viirya) [inline datafusion/sqllogictest/test_files/scalar.slt:1897]: The end-to-end tests are not failed, but I think it is good to have it to be clear.

For end-to-end execution, `LEFT('APACHE', 1)` is a foldable literal. So expression optimizer optimizes it to a scalar literal.

But if developers use the API `create_physical_expr` directly, the reported error will happen.

**Comment by viirya** [comment]
> thread 'sql::expr::test_regex_expressions' panicked at datafusion/core/tests/sql/mod.rs:500:38:
called `Result::unwrap()` on an `Err` value: ...

For the CI failure, it is an issue on how `try_from_array` handles NULL value in `ListArray`. Opened #7969 to fix it.

**Review by alamb — APPROVED** [review]
Thank you @viirya  -- I don't know about you but I found the code in `make_scalar_function_with_hints` very hard to follow as it went through several layers of indirection for reasons I didn't fully understand. 

Maybe this is something we can cleanup at some point -- maybe part of https://github.com/apache/arrow-datafusion/issues/7110  / https://github.com/apache/arrow-datafusion/pull/7752 that @2010YOUY01  has been working on


cc @suremarc

**Comment by viirya** [comment]
> Thank you @viirya -- I don't know about you but I found the code in `make_scalar_function_with_hints` very hard to follow as it went through several layers of indirection for reasons I didn't fully understand.

Yea, I agreed. We probably can simply it.

**Review by Weijun-H — APPROVED** [review]
LGTM! @viirya 👍

**Review by jayzhan211 — APPROVED** [review]
LGTM

**Comment by viirya** [comment]
Thank you @alamb @jayzhan211 @Weijun-H
